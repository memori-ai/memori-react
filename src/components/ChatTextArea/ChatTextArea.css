.memori-chat-textarea {
    position: relative;
    height: 40px;
    box-sizing: border-box;
    flex: 1 0 auto;
    margin-right: 0.5rem;
    /* margin-left: 0.5rem; */
}

.memori-chat-textarea--inner {
    position: absolute;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    overflow: hidden;
    height: 1000px;
    max-height: 40px;
    box-sizing: border-box;
    border: 1px solid lightgray;
    border-radius: 10px;
    background: #fff;
    transition: max-height 0.2s ease-in-out;
}

.memori-chat-textarea--expanded .memori-chat-textarea--inner {
    max-height: 200px;
}

.memori-chat-textarea--disabled .memori-chat-textarea--inner {
    background: #e3e3e3;
    opacity: 0.8;
}

textarea.memori-chat-textarea--input {
    width: 100%;
    height: 100%;
    box-sizing: border-box;
    padding: 6.5px 11px;
    border: none;
    background: #fff;
    color: #000;
    font-family: var(--memori-font-family);
    resize: none;
}

textarea.memori-chat-textarea--input:focus {
    outline: none;
}

.memori-chat-textarea--disabled textarea.memori-chat-textarea--input {
    cursor: not-allowed;
}

.memori-chat-textarea--expand {
    flex-grow: 0;
    margin: 0.25rem;
}

button.memori-chat-textarea--expand-button {
    padding: 0.25rem 0.5rem;
    color: #000;
    opacity: 0.6;
    transition: all 0.2s ease-in-out;
}

button.memori-chat-textarea--expand-button:hover,
button.memori-chat-textarea--expand-button:focus {
    color: #000;
    opacity: 1;
}

.memori-chat-textarea:not(.memori-chat-textarea--disabled) .memori-chat-textarea--inner:hover,
.memori-chat-textarea:not(.memori-chat-textarea--disabled) .memori-chat-textarea--inner:focus,
.memori-chat-textarea:not(.memori-chat-textarea--disabled) .memori-chat-textarea--inner:focus-within {
    border-color: var(--memori-primary);
    box-shadow: none;
    outline: none;
}

.memori-chat-textarea:not(.memori-chat-textarea--disabled)
    .memori-chat-textarea--inner:focus
    .memori-chat-textarea--expand-button,
.memori-chat-textarea:not(.memori-chat-textarea--disabled)
    .memori-chat-textarea--inner:focus-within
    .memori-chat-textarea--expand-button {
    opacity: 1;
}

.memori-chat-textarea:not(.memori-chat-textarea--disabled) .memori-chat-textarea-inner:focus-within {
    box-shadow: 0 0.2rem 0.33rem #b5b6c4;
}
