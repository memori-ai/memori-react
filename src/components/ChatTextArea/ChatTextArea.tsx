import React, { useState } from 'react';
import cx from 'classnames';
import Button from '../ui/Button';
import Expand from '../icons/Expand';
import FullscreenExit from '../icons/FullscreenExit';
import { useTranslation } from 'react-i18next';

export interface Props {
  disabled?: boolean;
  value: string;
  onChange: (value: string) => void;
  onPressEnter?: (e: React.KeyboardEvent<HTMLTextAreaElement>) => void;
  onFocus?: (e: React.FocusEvent) => void;
  onBlur?: (e: React.FocusEvent) => void;
}

const ChatTextArea: React.FC<Props> = ({
  disabled = false,
  value,
  onChange,
  onPressEnter,
  onFocus,
  onBlur,
}) => {
  const { t } = useTranslation();
  const [expanded, setExpanded] = useState(false);

  return (
    <div
      data-testid="chat-textarea"
      className={cx('memori-chat-textarea', {
        'memori-chat-textarea--expanded': expanded,
        'memori-chat-textarea--disabled': disabled,
      })}
    >
      <div className="memori-chat-textarea--inner">
        <textarea
          className="memori-chat-textarea--input"
          disabled={disabled}
          value={value}
          onChange={e => {
            onChange(e.target.value);
          }}
          onKeyDownCapture={e => {
            if (e.key === 'Enter' && !e.shiftKey && onPressEnter) {
              onPressEnter(e);
            }
          }}
          onFocus={onFocus}
          onBlur={onBlur}
        />
        <div className="memori-chat-textarea--expand">
          <Button
            className={cx('memori-chat-textarea--expand-button')}
            onClick={() => setExpanded(!expanded)}
            padded={false}
            ghost
            title={
              expanded ? t('collapse') || 'Collapse' : t('expand') || 'Expand'
            }
            icon={expanded ? <FullscreenExit /> : <Expand />}
          />
        </div>
      </div>
    </div>
  );
};

export default ChatTextArea;
