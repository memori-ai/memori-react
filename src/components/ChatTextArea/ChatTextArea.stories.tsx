import React from 'react';
import { Meta, Story } from '@storybook/react';
import ChatTextArea, { Props } from './ChatTextArea';
import I18nWrapper from '../../I18nWrapper';
import './ChatTextArea.css';

const meta: Meta = {
  title: 'Widget/Chat textarea',
  component: ChatTextArea,
  argTypes: {
    disabled: {
      control: {
        type: 'boolean',
      },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<Props> = args => {
  const [userMessage, setUserMessage] = React.useState(args.value);

  return (
    <I18nWrapper>
      <div style={{ paddingTop: '10rem' }}>
        <ChatTextArea {...args} value={userMessage} onChange={setUserMessage} />
      </div>
    </I18nWrapper>
  );
};

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  value: '',
  onChange: () => {},
};

export const WithValue = Template.bind({});
WithValue.args = {
  value: 'Suspendisse sit amet volutpat velit.',
  onChange: () => {},
};

export const WithLongText = Template.bind({});
WithLongText.args = {
  value:
    'Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur. Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur.',
  onChange: () => {},
};
