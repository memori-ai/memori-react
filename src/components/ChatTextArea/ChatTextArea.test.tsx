import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import ChatTextArea from './ChatTextArea';

it('renders ChatTextArea unchanged', () => {
  const { container } = render(<ChatTextArea value="" onChange={jest.fn()} />);
  expect(container).toMatchSnapshot();
});

it('renders ChatTextArea with value unchanged', () => {
  const { container } = render(
    <ChatTextArea
      value="Proin libero ante, dignissim sit amet turpis a, pretium condimentum dolor."
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders ChatTextArea with long text unchanged', () => {
  const { container } = render(
    <ChatTextArea
      value="Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur. Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur."
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders ChatTextArea expanded unchanged', () => {
  render(
    <ChatTextArea
      value="Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur. Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur."
      onChange={jest.fn()}
    />
  );

  fireEvent.click(screen.getByRole('button'));

  expect(screen.getByTestId('chat-textarea').classList).toContain(
    'memori-chat-textarea--expanded'
  );
});
