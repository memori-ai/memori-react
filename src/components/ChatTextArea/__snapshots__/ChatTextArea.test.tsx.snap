// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ChatTextArea unchanged 1`] = `
<div>
  <div
    class="memori-chat-textarea"
    data-testid="chat-textarea"
  >
    <div
      class="memori-chat-textarea--inner"
    >
      <textarea
        class="memori-chat-textarea--input"
      />
      <div
        class="memori-chat-textarea--expand"
      >
        <button
          class="memori-button memori-button--ghost memori-button--rounded memori-button--icon-only memori-chat-textarea--expand-button"
          title="expand"
        >
          <span
            class="memori-button--icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M855 160.1l-189.2 23.5c-6.6.8-9.3 8.8-4.7 13.5l54.7 54.7-153.5 153.5a8.03 8.03 0 0 0 0 11.3l45.1 45.1c3.1 3.1 8.2 3.1 11.3 0l153.6-153.6 54.7 54.7a7.94 7.94 0 0 0 13.5-4.7L863.9 169a7.9 7.9 0 0 0-8.9-8.9zM416.6 562.3a8.03 8.03 0 0 0-11.3 0L251.8 715.9l-54.7-54.7a7.94 7.94 0 0 0-13.5 4.7L160.1 855c-.6 5.2 3.7 9.5 8.9 8.9l189.2-23.5c6.6-.8 9.3-8.8 4.7-13.5l-54.7-54.7 153.6-153.6c3.1-3.1 3.1-8.2 0-11.3l-45.2-45z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders ChatTextArea with long text unchanged 1`] = `
<div>
  <div
    class="memori-chat-textarea"
    data-testid="chat-textarea"
  >
    <div
      class="memori-chat-textarea--inner"
    >
      <textarea
        class="memori-chat-textarea--input"
      >
        Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur. Suspendisse sit amet volutpat velit. Nunc at commodo tortor, id rutrum nunc. Vivamus condimentum vel nunc et congue. Ut laoreet imperdiet nisi ac finibus. Suspendisse molestie risus a justo sagittis efficitur.
      </textarea>
      <div
        class="memori-chat-textarea--expand"
      >
        <button
          class="memori-button memori-button--ghost memori-button--rounded memori-button--icon-only memori-chat-textarea--expand-button"
          title="expand"
        >
          <span
            class="memori-button--icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M855 160.1l-189.2 23.5c-6.6.8-9.3 8.8-4.7 13.5l54.7 54.7-153.5 153.5a8.03 8.03 0 0 0 0 11.3l45.1 45.1c3.1 3.1 8.2 3.1 11.3 0l153.6-153.6 54.7 54.7a7.94 7.94 0 0 0 13.5-4.7L863.9 169a7.9 7.9 0 0 0-8.9-8.9zM416.6 562.3a8.03 8.03 0 0 0-11.3 0L251.8 715.9l-54.7-54.7a7.94 7.94 0 0 0-13.5 4.7L160.1 855c-.6 5.2 3.7 9.5 8.9 8.9l189.2-23.5c6.6-.8 9.3-8.8 4.7-13.5l-54.7-54.7 153.6-153.6c3.1-3.1 3.1-8.2 0-11.3l-45.2-45z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders ChatTextArea with value unchanged 1`] = `
<div>
  <div
    class="memori-chat-textarea"
    data-testid="chat-textarea"
  >
    <div
      class="memori-chat-textarea--inner"
    >
      <textarea
        class="memori-chat-textarea--input"
      >
        Proin libero ante, dignissim sit amet turpis a, pretium condimentum dolor.
      </textarea>
      <div
        class="memori-chat-textarea--expand"
      >
        <button
          class="memori-button memori-button--ghost memori-button--rounded memori-button--icon-only memori-chat-textarea--expand-button"
          title="expand"
        >
          <span
            class="memori-button--icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M855 160.1l-189.2 23.5c-6.6.8-9.3 8.8-4.7 13.5l54.7 54.7-153.5 153.5a8.03 8.03 0 0 0 0 11.3l45.1 45.1c3.1 3.1 8.2 3.1 11.3 0l153.6-153.6 54.7 54.7a7.94 7.94 0 0 0 13.5-4.7L863.9 169a7.9 7.9 0 0 0-8.9-8.9zM416.6 562.3a8.03 8.03 0 0 0-11.3 0L251.8 715.9l-54.7-54.7a7.94 7.94 0 0 0-13.5 4.7L160.1 855c-.6 5.2 3.7 9.5 8.9 8.9l189.2-23.5c6.6-.8 9.3-8.8 4.7-13.5l-54.7-54.7 153.6-153.6c3.1-3.1 3.1-8.2 0-11.3l-45.2-45z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
